src/util/address.rs

```diff
@@ -655,6 +666,11 @@ impl FromStr for Address {
                 return Err(Error::InvalidSegwitV0ProgramLength(program.len()));
             }
 
+            // Encoding check
+            if version.bech32_variant() != variant {
+                return Err(Error::InvalidBech32Variant);
```

Comment:
nit: This could provide more context about the error by also providing version and variant.

Reply:
Fixed in 6848ea43aa53e903323408bfe78bec342dc9efa4